spark/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala

```diff
@@ -583,6 +588,191 @@ trait DataSkippingReaderBase
+      // The expression is an eligible reference to an attribute.
+      // Do NOT allow partition-like filtering on timestamp columns because timestamps are truncated
+      // to millisecond precision, meaning that we can't guarantee that the collected minVal and
+      // maxVal are the same.
+      // Applying these partition-like filters will generally only be beneficial if a large
+      // percentage of files have the same min-max value. As a rough heuristic, only allow rewriting
+      // expressions that reference only the clustering columns (since these columns are more likely
+      // to have the same min-max values).
+      case SkippingEligibleColumn(c, SkippingEligibleDataType(dt))
+        if dt != TimestampType && dt != TimestampNTZType &&
+          clusteringColumnPaths.exists(SchemaUtils.areLogicalNamesEqual(_, c.reverse)) =>
+        // Only rewrite the expression if all stats are collected for this column.
+        val minStatsCol = StatsColumn(MIN, c, dt)
+        val maxStatsCol = StatsColumn(MAX, c, dt)
+        val nullCountStatsCol = StatsColumn(NULL_COUNT, c, dt)
+        for {
+          minCol <- getStatsColumnOpt(minStatsCol);
+          maxCol <- getStatsColumnOpt(maxStatsCol);
+          nullCol <- getStatsColumnOpt(nullCountStatsCol)
+        } yield {
+          val resolvedAttribute = ResolvedPartitionLikeReference(
+            Seq(minStatsCol, maxStatsCol, nullCountStatsCol),
+            minCol.expr,
+            maxCol.expr,
+            nullCol.expr)
+          (minCol.expr, Set(resolvedAttribute))
+        }
+      // For other attribute references, we can't safely rewrite the expression.
+      case SkippingEligibleColumn(_, _) => None
+      // Don't attempt data skipping on a nondeterministic expression, since the value returned
+      // might be different when executed twice on the same input.
+      // For example, rand() > 0.5 would return ~25% of records if used in data skipping, while the
+      // user would expect ~50% of records to be returned.
+      case other if !other.deterministic => None
+      // Inline subquery results to support InSet. The subquery should generally have already been
+      // evaluated.
+      case in: InSubqueryExec =>
+        // Values may not be defined if the subquery has been skipped - we can't apply this filter.
+        in.values().flatMap { possiblyNullValues =>
+          // Rewrite the children of InSubqueryExec, then replace the subquery with an InSet
+          // containing the materialized values.
+          rewriteDataFiltersAsPartitionLikeInternal(in.child, clusteringColumnPaths).flatMap {
+            case (rewrittenChildren, referencedStats) =>
+              Some(InSet(rewrittenChildren, possiblyNullValues.toSet), referencedStats)
+          }
+        }
+      // Don't allow rewriting UDFs - even if deterministic, UDFs might have some unexpected
+      // side effects when executed twice.
+      case _: UserDefinedExpression => None
```

Comment:
Is the problem that spark UDF are marked "nondeterministic" by default, and their creator has to recognize whether that's true and override as needed? So basically we have to assume that any UDF which claims "nondeterministic" is lying?

Reply:
Exactly - we can't trust that UDFs are deterministic (as the creators rarely, if ever, recognize that UDFs are nondeterministic).

The other issue is that UDFs can be arbitrarily slow/expensive to call, so we risk a performance regression when double executing UDFs (some users are also sensitive to double executing UDFs due to having UDFs that call a metered API).